### ISSUE TITLE
`--compare-branch` should default to `@{upstream}` and not `origin/master`

### ISSUE THREAD

**Issue by ydirson** [Bachmann1234/diff_cover#141]
Not all branches are forked off a branch named `master` in remote `origin`.  When an upstream has been configured for current branch it should be the first choice; when there is none `origin/master` is still the obvious default.

**Comment by jaraco** [comment]
Additionally, the ecosystem is [transitioning](https://github.blog/changelog/2020-10-01-the-default-branch-for-newly-created-repositories-is-now-main/) from `master` to `main` as the default branch name, so `origin/master` is no longer an obvious default. I'd recommend to infer the upstream branch. I'd even go as far as to say that `main` should be the preferred fallback, but if any of `main`, `master`, `default`, or possibly others are present, then they will be preferred (in that order) if no upstream is indicated.

**Comment by Bachmann1234** [comment]
A PR that simply changes the default from `master` to `main` would likely break existing users which I don't really want to do.

However, a PR that works by detecting the upstream mainline branch would be more than acceptable for merging.

**Comment by ssbarnea** [comment]
Oops it seems quite sad that after more than two years diff-cover is still not able to recognize default branches that are not "master". 

As someone trying diff-cover for the first time that exceptions become a road blocker.
```
  File "/Users/ssbarnea/.pyenv/versions/3.9.5/lib/python3.9/site-packages/diff_cover/git_diff.py", line 67, in diff_committed
    return execute(self._default_git_args + self._default_diff_args + [diff_range])[
  File "/Users/ssbarnea/.pyenv/versions/3.9.5/lib/python3.9/site-packages/diff_cover/command_runner.py", line 42, in execute
    raise CommandError(stderr)
diff_cover.command_runner.CommandError: fatal: ambiguous argument 'origin/master...HEAD': unknown revision or path not in the working tree.
Use '--' to separate paths from revisions, like this:
'git <command> [<revision>...] -- [<file>...]'

FAIL: 1
```

**Comment by Bachmann1234** [comment]
I agree it's frustrating. But I can't simply change the default without breaking existing users workflows

And while githubs default is `main` gitlab and git itself still very much default to `master`. I think `main` is a better name but it's not relevant to this project.

I'm the maintainer of this project. As such I try and provide timely feedback on prs. If I get a bug report with enough info to act on I reproduce it in a test and fix it. Whenever a fix or a feature comes in I release it asap.

But I typically don't implement new functionality. Someone has to care enough to implement it themselves. 

In this case I'll take a look, probably this weekend. I might just resort to providing a more helpful error message or perhaps trying `main` if `master` does not exist. As far as I know there is not really a clean way of detecting the "main" branch.

**Comment by kasium** [comment]
My 2 cents here: I couldn't find a option as well to find the default branch. Maybe a CLI option would help the most. Else we could check if master exists and if not fall back to main (no clue if there are projects with both).

**Comment by ydirson** [comment]
Well, is there anything wrong with using `@{upstream}` as I suggested in the title ?

**Comment by Bachmann1234** [comment]
@ydirson unless I'm mistaken @{upstream} is just the upstream version of your branch not the "core branch that represents the production ready code". 

So I can have a branch "feature". Upstream would just be the version of that branch that is "upstream" to me. Usually github but technically could just be another version of the repo since git is designed to be fully decentralized

**Comment by Bachmann1234** [comment]
Ok, so I did some more reading and looked over the code and decided I was wrong about a few things.

1. gitlab has switched over to main
2. As y'all were saying git is actively working to change over to main as a default. I knew they added a config value but I had underestimated how far along this plan was

Sorry about that, most of the time when looking at diff_cover issues its either at night when im already a bit fried from the day or its weekend. 

So I looked over the code and I decided rather than "trying both main and master" to just bite the bullet and make the breaking change. 

https://github.com/Bachmann1234/diff_cover/pull/212

This PR will switch main as the default. It also includes a helpful error message when the main branch is not found (Something that should have been there from the start....) 

Im gonna let that pr sit there for a little bit (@kasium has been quite active lately so maybe ill get their thoughts). Assuming no major issues ill probably merge and release it either sunday or early next week. 

Im _probably_ gonna follow that up with a PR to change this repo to have main as its default branch.
